joomla/joomla-cms
# [5.1] Change redirect url (from headers['Location']) from array to string in CurlTransport

The problem was discovered when using HTTP Factory for an external request to the server that performs redirection. The `request` method executes a request to the specified url of the `string` type. If the response code indicates a redirect, then the redirect url is obtained from the `Location` header. After that, another request is executed with a new url. 

![image](https://github.com/joomla/joomla-cms/assets/6236403/9ade336a-4ee7-423d-9dfd-9caba7803855)

The redirect url is an instance of the `Uri` class. During the operation of this class, the `UrlHelper` class is called, in which the url is parsed. When making the first request, the url type is a **string**. However, in the case of redirection, **the url is an array with one element**, since `$response->headers['Location']` is passed immediately as an argument. Therefore, when trying to make a second request for a new URl, the error **parse_url() occurs: Argument #1 ($url) must be of type string, array given** in the `UriHelper` class line 38. 

![image](https://github.com/joomla/joomla-cms/assets/6236403/a5f6cd8f-f2bf-4a69-be98-8bd6870b8ea3)

This screenshot shows that the content of the Location header is an array, not a string. Which is the reason for the error.
![image](https://github.com/joomla/joomla-cms/assets/6236403/a57d9359-e4c3-42ef-bfd2-6dbbac6124e7)

The problem was found only when using the `CURLOPT_FOLLOWLOCATION = false` parameter. In this case, Joomla executes the redirect url request on its own.

### Summary of Changes
Change a `$redirect_uri = new Uri($response->headers['Location']);` to `$redirect_uri = new Uri($response->headers['Location'][0]);`

### Testing Instructions
It may not be the most correct, but the fastest way to check the result of this change is to run the query directly from the template `index.php` file.

1. Open it index.php Your frontend template
2. Add the following code to it
```php
use Joomla\CMS\Http\HttpFactory;

$options = new Registry();
$options->set(
	'transport.curl',
	[
		CURLOPT_FOLLOWLOCATION   => false, 
	]
);

$http = (new HttpFactory)->getHttp($options, ['curl', 'stream']);

$url = 'enter here any URL with redirect ';

$response = $http->get($url);
echo '<pre>';
print_r($response->body);
echo '</pre>';

```
3. Run the request (refresh the page) and you will see an error `parse_url(): Argument #1 ($url) must be of type string, array given` 
4.  Apply this PR (download the build and update Joomla 5.0.2) 

![image](https://github.com/joomla/joomla-cms/assets/6236403/f8416463-d66e-49b5-87c2-2c9bce252a95)

or in the file `libraries/src/Http/Transport/CurlTransport.php` find the line `$redirect_uri = new Uri($response->headers['Location'])` and replace it to `$redirect_uri = new Uri($response->headers['Location'][0]);`
5. Run the request again (refresh the page again)
6. Error has gone

### Actual result BEFORE applying this Pull Request
An error `parse_url(): Argument #1 ($url) must be of type string, array given` in ROOT\libraries\vendor\joomla\uri\src\UriHelper.php:38

![image](https://github.com/joomla/joomla-cms/assets/6236403/53409480-dfa6-43c5-b2a6-3c76ed3bcef2)


### Expected result AFTER applying this Pull Request
Error has gone


### Link to documentations
Please select:
